### PR TITLE
Ascii.Equals and Ascii.EqualsIgnoreCase

### LINKED ISSUE #84887: Ascii.Equals
### Background and motivation

Based on an ASP.NET and dotnet/runtime code [study](https://gist.github.com/adamsitnik/19081bf757801b2e91990f90a60086fe), I concluded that from all of the methods that were initially proposed in #28230 and were reverted in https://github.com/dotnet/runtime/pull/75012/commits/0d69abd9cfdc0c9988f81743301e064f237b97a5 before #75012 got merged, we need only `Equals` and `EqualsIgnoreCase`.

The tricky part was deciding what should happen when both input values are equal, but they are not valid ASCII. Example:

```cs
Ascii.Equals(new byte[] { 128 }, new char[] { (char)128 });
```

The initial proposal (#28230) suggested to throw. We have received a very strong push back about it. Based on real-life use cases from dotnet/runtime and ASP.NET:

https://github.com/dotnet/aspnetcore/blob/8968058c9e5fdfdd1242426a03dc80609997edab/src/Http/Routing/src/Matching/Ascii.cs#L16
https://github.com/dotnet/aspnetcore/blob/8968058c9e5fdfdd1242426a03dc80609997edab/src/Shared/ServerInfrastructure/StringUtilities.cs#L420

together with @stephentoub and @grabyourpitchforks we came to agreement, that these methods should be used only when at least one of the input buffers is guaranteed to be valid ASCII (common use case: comparing user input with a const value) and the equality checks should return `false` in such case.

### API Proposal

```cs
namespace System.Text;

public static partial class Ascii
{
    /// <summary>
    /// Determines whether the provided buffers contain equal ASCII characters.
    /// </summary>
    /// <param name="left">The buffer to compare with <paramref name="right" />.</param>
    /// <param name="right">The buffer to compare with <paramref name="left" />.</param>
    /// <returns><see langword="true" /> if the corresponding elements in <paramref name="left" /> and <paramref name="right" /> were equal and ASCII. <see langword="false" /> otherwise.</returns>
    /// <remarks>If both buffers contain equal, but non-ASCII characters, the method returns <see langword="false" />.</remarks>
    public static bool Equals(ReadOnlySpan<byte> left, ReadOnlySpan<byte> right);
    public static bool Equals(ReadOnlySpan<byte> left, ReadOnlySpan<char> right);
    public static bool Equals(ReadOnlySpan<char> left, ReadOnlySpan<char> right);
    
    /// <summary>
    /// Determines whether the provided buffers contain equal ASCII characters, ignoring case considerations.
    /// </summary>
    /// <param name="left">The buffer to compare with <paramref name="right" />.</param>
    /// <param name="right">The buffer to compare with <paramref name="left" />.</param>
    /// <returns><see langword="true" /> if the corresponding elements in <paramref name="left" /> and <paramref name="right" /> were equal ignoring case considerations and ASCII. <see langword="false" /> otherwise.</returns>
    /// <remarks>If both buffers contain equal, but non-ASCII characters, the method returns <see langword="false" />.</remarks>
    public static bool EqualsIgnoreCase(ReadOnlySpan<byte> left, ReadOnlySpan<byte> right);
    public static bool EqualsIgnoreCase(ReadOnlySpan<byte> left, ReadOnlySpan<char> right);
    public static bool EqualsIgnoreCase(ReadOnlySpan<char> left, ReadOnlySpan<char> right);
}
```

### Alternative Designs

`Equals(ReadOnlySpan<byte> left, ReadOnlySpan<byte> right)` and `Equals(ReadOnlySpan<char> left, ReadOnlySpan<char> right)` could be removed, as they can be expressed by calling `left.SequenceEqual(right)` with `Ascii.IsValid` calls if needed.

### PR DESCRIPTION
**Pull request by adamsitnik**
fixes #84887

### CODE REVIEW

File: src/libraries/System.Net.Http/src/System/Net/Http/ByteArrayHelpers.cs
Comment (MihaZupan) [inline src/libraries/System.Net.Http/src/System/Net/Http/ByteArrayHelpers.cs:9]: Can you please delete this helper class entirely? There are only ~5 places that call into this which can now use the Ascii API directly.

Comment (adamsitnik) [reply src/libraries/System.Net.Http/src/System/Net/Http/ByteArrayHelpers.cs:9]: My goal was to add the debug asserts to verify that at least one of the inputs is always ASCII. But the CI is green, so it's confirmed and I can remove it now.

Comment (stephentoub) [reply src/libraries/System.Net.Http/src/System/Net/Http/ByteArrayHelpers.cs]: These APIs may be used where neither input is fully ascii and they need to account for that, regardless, of course. I believe we agreed they could simply return false in such cases (though wouldn't be guaranteed to).

File: src/libraries/System.Text.Encoding/tests/Ascii/EqualsTests.cs
Comment (jeffhandley) [inline src/libraries/System.Text.Encoding/tests/Ascii/EqualsTests.cs:66]: I'm generally not a fan of tests that validate many different things--in this case, the behavior of 6 different methods/overloads. With a single test here, only the first failure will get reported instead of the test results showing all of the failing asserts.

I recommend having 6 different `Theory` methods here instead--one for each method/overload.

File: src/libraries/System.Text.Encoding/tests/Ascii/EqualsTests.cs
Comment (jeffhandley) [inline src/libraries/System.Text.Encoding/tests/Ascii/EqualsTests.cs:123]: I also recommend 3 different `[Theory]`s here

File: src/libraries/System.Text.Encoding/tests/Ascii/EqualsTests.cs
Comment (jeffhandley) [inline src/libraries/System.Text.Encoding/tests/Ascii/EqualsTests.cs:99]: I also recommend 6 different `[Theory]`s here

File: src/libraries/System.Text.Encoding/tests/Ascii/EqualsTests.cs
Comment (jeffhandley) [inline src/libraries/System.Text.Encoding/tests/Ascii/EqualsTests.cs:10]: I _suggest_ adding a could tests that show real-world use cases like what we see in networking or aspnetcore, comparing against a constant. Such tests could use a non-ascii utf8 sequence that could match something provided by client input in one of those scenarios.

File: src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs
Comment (gfoidl) [inline src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs:309]:
```suggestion
                if (!UnicodeUtility.IsAsciiCodePoint(valueA | valueB))
```
Should work too and saves a branch. The MSB will only be set if one of the values is > 0x7F (non-ASCII).

File: src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs
Comment (gfoidl) [inline src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs:304]: Here is no vectorization needed?

File: src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs
Comment (gfoidl) [inline src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs:343]: IIRC `Vector64` uses the software fallback on x64.
[crossgen2 with main-branch](https://godbolt.org/z/65bxWG8v9) shows that too.

I don't know if there are plans or WIP to hw-accelerate Vector64 in the .NET 8 timeframe.
Otherwise maybe read the value as long into a Vector128 and do the widening then. Something like
```c#
static Vector128<ushort> LoadAndWiden(ref byte ptr)
{
    if (AdvSimd.IsSupported)
    {
        Vector64<byte> vec64 = Vector64.LoadUnsafe(ref ptr);
        return AdvSimd.ZeroExtendWideningLower(vec64);
    }
    else if (Sse2.IsSupported)
    {
        ulong value = Unsafe.ReadUnaligned<ulong>(ref ptr);
        Vector128<byte> vec = Vector128.CreateScalarUnsafe(value).AsByte();
        return Sse2.UnpackLow(vec, Vector128<byte>.Zero).AsUInt16();
    }
    else
    {
        Vector64<byte> vec64 = Vector64.LoadUnsafe(ref ptr);
        (Vector64<ushort> lower, Vector64<ushort> upper) = Vector64.Widen(vec64);
        return Vector128.Create(lower, upper);
    }
}
```

File: src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs
Comment (gfoidl) [inline src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs:356]: Same concern here for Vector64. Pass in `ulong` as `Utf8Utility.AllBytesInUInt64AreAscii` expects that.
So get rid of Vector64 where possible and use `ulong` instead?

**Comment by adamsitnik** [comment]
@gfoidl thank you for your suggestions and ideas! I won't have the time to implement all perf improvements now (I started paternity leave very recently and now I just want to get this issue off my TODO list). Would you be interested in sending a PR with improvements once this PR is merged?

**Comment by gfoidl** [comment]
> Would you be interested in sending a PR with improvements once this PR is merged?

Yep, I can do the follow up.

> paternity leave

Congratulations!

File: src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs
Comment (jeffhandley) [reply src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs:34]: Since we've already confirmed `left.Length == right.Length` above, I think we're OK without another bounds check here.

File: src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs
Comment (jeffhandley) [reply src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs:312]: Perhaps vectorization here would be a further perf opportunity that was missed. I recommend we proceed with mering this without it and potentially follow up with vectorizing it in a separate PR.

**Review by jeffhandley — APPROVED** [review]
Looks good to me. @adamsitnik requested we merge this without the vectorization improvements recommended by @gfoidl so that we can get in ahead of the Preview 4 snap (and while Adam is unavailable), and then potentially follow up with an additional PR to improve the vectorization.

@gfoidl -- would you like to submit a follow-up PR with those recommendations you made?

**Comment by gfoidl** [comment]
@jeffhandley yes I'll do a follow up PR to address the perf-points (maybe soon, otherwise in about 2 weeks as rough timeframe).
Bring this PR in...

**Comment by jeffhandley** [comment]
Awesome; thank you! I'll merge when green.

**Comment by gfoidl** [comment]
@jeffhandley follow up PR is here: https://github.com/dotnet/runtime/pull/85926

**Pull request merged**
